File: solutions/python/list-ops/1/list_ops.py
"""Implement basic list operations."""

# pylint: disable=redefined-builtin


def append(list1, list2):
    """Return list1 followed by list2."""
    return foldl(_append_item, list2, list(list1))


def concat(lists):
    """Flatten one level of nested lists."""
    return foldl(append, lists, [])


def filter(function, items):
    """Return items for which function returns true."""
    def keep_if(result, item):
        return _append_item(result, item) if function(item) else result

    return foldl(keep_if, items, [])


def length(items):
    """Return the number of items."""
    count = 0

    for _ in items:
        count += 1

    return count


def map(function, items):
    """Return the result of applying function to each item."""
    def append_mapped(result, item):
        result.append(function(item))
        return result

    return foldl(append_mapped, items, [])


def foldl(function, items, initial):
    """Fold items into the accumulator from the left."""
    result = initial

    for item in items:
        result = function(result, item)

    return result


def foldr(function, items, initial):
    """Fold items into the accumulator from the right."""
    return foldl(function, reverse(items), initial)


def reverse(items):
    """Return items in reverse order."""
    def prepend(result, item):
        result.insert(0, item)
        return result

    return foldl(prepend, items, [])


def _append_item(result, item):
    result.append(item)
    return result

File: solutions/fsharp/list-ops/1/ListOps.fs
module ListOps

let rec foldl folder state list =
    match list with
    | [] -> state
    | head :: tail -> foldl folder (folder state head) tail

let rec foldr folder state list =
    match list with
    | [] -> state
    | head :: tail -> folder (foldr folder state tail) head

let length list =
    foldl (fun count _ -> count + 1) 0 list

let reverse list =
    foldl (fun reversed head -> head :: reversed) [] list

let map f list =
    foldr (fun mapped head -> f head :: mapped) [] list

let filter predicate list =
    foldr (fun filtered head -> if predicate head then head :: filtered else filtered) [] list

let append xs ys =
    foldr (fun appended head -> head :: appended) ys xs

let concat lists =
    foldr (fun concatenated current -> append current concatenated) [] lists
